compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETileAndFuse.cpp

```diff
@@ -0,0 +1,337 @@
+  OpBuilder::InsertionGuard g(rewriter);
+  while (!candidates.empty()) {
+    // Traverse the slices in BFS fashion.
+    tensor::ExtractSliceOp candidateSliceOp = candidates.front();
+    candidates.pop_front();
+
+    // Materialize the slice of the producer in place.
+    std::optional<scf::SCFFuseProducerOfSliceResult> fusedProducer =
+        tileAndFuseProducerOfSlice(rewriter, candidateSliceOp, forallLoop);
+    if (!fusedProducer) continue;
+
+    // Add more fusion candidates to the worklist.
+    for (auto tiledOp : fusedProducer->tiledOps) {
+      addCandidateSlices(tiledOp, candidates);
+      tiledOps.push_back(tiledOp);
+    }
+  }
+
+  for (auto [index, origVal] : llvm::enumerate(yieldedValuesToOrigValues)) {
+    Value replacement = forallLoop.getResult(index);
+    rewriter.replaceUsesWithIf(origVal, replacement, [&](OpOperand &use) {
+      return !isIgnoredUser(use.getOwner()) &&
+             dominanceInfo.properlyDominates(forallLoop, use.getOwner());
+    });
+  }
+
+  return success();
+}
+
+/// This pass starts with the last TilingInterface operation, tiles the op and
+/// fuses its producers recursively. The `tilingLevel` must be specified. It
+/// picks the `tilingLevel`-th list as tiling sizes from lowering_config.
+class AMDAIETileAndFusePass
+    : public AMDAIETileAndFuseBase<AMDAIETileAndFusePass> {
+ public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<gpu::GPUDialect, scf::SCFDialect>();
+  }
+
+  AMDAIETileAndFusePass() = default;
+  AMDAIETileAndFusePass(int64_t tilingLevel = -1) {
+    this->tilingLevel.setValue(tilingLevel);
+  }
+  AMDAIETileAndFusePass(const AMDAIETileAndFusePass &pass){};
+
+  void runOnOperation() override;
+};
+
+void AMDAIETileAndFusePass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  auto funcOp = getOperation();
+
+  TilingInterface consumerOp;
+  funcOp->walk<WalkOrder::PostOrder, ReverseIterator>([&](TilingInterface op) {
+    // Find the next consumer op if it does not have loops.
+    if (op.getLoopIteratorTypes().empty()) return WalkResult::advance();
+    consumerOp = op;
+    return WalkResult::interrupt();
+  });
+  if (!consumerOp) {
+    LLVM_DEBUG(llvm::dbgs() << "----- skip, no consumer op -----\n");
+    return;
+  }
+
+  LLVM_DEBUG(llvm::dbgs() << "consumerOp: " << consumerOp << "\n");
+  LLVM_DEBUG(llvm::dbgs() << "tilingLevel: " << tilingLevel << "\n");
+  // TODO: Currently hardcoding the tile size to make the tile and fuse run for
+  //       at least the first level. So, we can generalize this by adding tile
+  //       level, similar to LLVMCPUTileAndFuse.cpp.
+  SmallVector<OpFoldResult> tileSizes = getAsIndexOpFoldResult(context, {8, 8});
+  auto options = scf::SCFTilingOptions().setTileSizes(tileSizes);
+  options.setMapping(
+      {gpu::GPUBlockMappingAttr::get(context, gpu::MappingId::DimY),
+       gpu::GPUBlockMappingAttr::get(context, gpu::MappingId::DimX)});
```

Comment:
This might be ok for now, but ideally we'll need some AIE specific attributes similar to the gpu ones.

Reply:
Yep, you're correct. We'll be revisiting this all iteratively. Currently we're trying to mimic exactly what the [pad transform dialect script gets us](https://github.com/nod-ai/iree-amd-aie/blob/1e2aa24c856ce33fcef3d749e98c4ac2aff407ac/tests/samples/matmul_fill_spec_pad.mlir#L39). Similar thing for the pack transform dialect script, for now.